File: java/space-age/src/main/java/SpaceAge.java
public class SpaceAge {

   private enum Period {

       EARTH(31557600.0),
       MERCURY((long)(0.2408467 * 31557600.0)),
       VENUS((long)(0.61519726 * 31557600.0)),
       MARS((long)(1.8808158 * 31557600.0)),
       JUPITER((long)(11.862615 * 31557600.0)),
       SATURN((long)(29.447498 * 31557600.0)),
       URANUS((long)(84.016846 * 31557600.0)),
       NEPTUNE((long)(164.79132 * 31557600.0));

       private double seconds;

       Period(double seconds) {
            this.seconds = seconds;
       }

       double getSeconds() {
           return this.seconds;
       }
   }

    private long seconds;

    public SpaceAge(long seconds) {
        this.seconds = seconds;
    }

    public long getSeconds() {
        return seconds;
    }

    public double onEarth() {
        return (double)this.seconds / Period.EARTH.getSeconds();
    }

    public double onMercury() {
        return (double)this.seconds / Period.MERCURY.getSeconds();
    }

    public double onVenus() {
        return (double)this.seconds / Period.VENUS.getSeconds();
    }

    public double onMars() {
        return (double)this.seconds / Period.MARS.getSeconds();
    }

    public double onJupiter() {
        return (double)this.seconds / Period.JUPITER.getSeconds();
    }

    public double onSaturn() {
        return (double)this.seconds / Period.SATURN.getSeconds();
    }

    public double onUranus() {
        return (double)this.seconds / Period.URANUS.getSeconds();
    }

    public double onNeptune() {
        return (double)this.seconds / Period.NEPTUNE.getSeconds();
    }
}

File: ruby/space-age/space_age.rb
class SpaceAge

  EARTH_PERIOD_IN_SECONDS = 31557600
  PLANET_PERIODS = {
    :earth => 1.0,
    :mercury => 0.2408467,
    :venus => 0.61519726,
    :mars => 1.8808158,
    :jupiter => 11.862615,
    :saturn => 29.447498,
    :uranus => 84.016846,
    :neptune => 164.79132
  }

  attr_reader :seconds

  def initialize(seconds)
    @seconds = seconds
  end

  PLANET_PERIODS.keys.each do |planet|
    define_method "on_#{planet}" do
      send(:on, planet)
    end
  end

  def on(planet)
    @seconds / PLANET_PERIODS.fetch(planet,1) / EARTH_PERIOD_IN_SECONDS
  end

end
